src/time_sequence.hpp

```diff
@@ -0,0 +1,251 @@
+    //
+
+    time_seq(): time_seq(dummy_seq()) {}
+
+    template <
+        typename Impl,
+        typename = typename util::enable_if_t<
+            !std::is_same<typename std::decay<Impl>::type,
+                          time_seq>::value>>
+    time_seq(Impl&& impl):
+        impl_(new wrap<Impl>(std::forward<Impl>(impl)))
+    {}
+
+    time_seq(time_seq&& other) = default;
+    time_seq& operator=(time_seq&& other) = default;
+
+    time_seq(const time_seq& other):
+        impl_(other.impl_->clone())
+    {}
+
+    time_seq& operator=(const time_seq& other) {
+        impl_ = other.impl_->clone();
+        return *this;
+    }
+
+    //
+    // time sequence interface
+    //
+
+    // Get the current time in the stream.
+    // Does not modify the state of the stream, i.e. multiple calls to
+    // next() will return the same time in the absence of calls to pop(),
+    // advance() or reset().
+    time_type next() {
+        return impl_->next();
+    }
+
+    // Move the generator to the next time in the stream.
+    void pop() {
+        impl_->pop();
+    }
+
+    // Reset the generator to the same state that it had on construction.
+    void reset() {
+        impl_->reset();
+    }
+
+    // Update state of the generator such that the time returned by next() is
+    // the first time with delivery time >= t.
+    void advance(time_type t) {
+        return impl_->advance(t);
+    }
+
+private:
+    struct interface {
+        virtual time_type next() = 0;
+        virtual void pop() = 0;
+        virtual void advance(time_type t) = 0;
+        virtual void reset() = 0;
+        virtual std::unique_ptr<interface> clone() = 0;
+        virtual ~interface() {}
+    };
+
+    std::unique_ptr<interface> impl_;
+
+    template <typename Impl>
+    struct wrap: interface {
+        explicit wrap(const Impl& impl): wrapped(impl) {}
+        explicit wrap(Impl&& impl): wrapped(std::move(impl)) {}
+
+        time_type next() override {
+            return wrapped.next();
+        }
+        void pop() override {
+            return wrapped.pop();
+        }
+        void advance(time_type t) override {
+            return wrapped.advance(t);
+        }
+        void reset() override {
+            wrapped.reset();
+        }
+        std::unique_ptr<interface> clone() override {
+            return std::unique_ptr<interface>(new wrap<Impl>(wrapped));
+        }
+
+        Impl wrapped;
+    };
+
+    struct dummy_seq {
+        time_type next() { return max_time; }
+        void pop() {}
+        void reset() {}
+        void advance(time_type t) {};
+    };
+
+};
+
+// Sequence of time points prescribed by a vector
+struct vector_time_seq {
+    vector_time_seq(std::vector<time_type> seq):
+        seq_(std::move(seq))
+    {
+        // Ensure that the time values are sorted.
+        if(!std::is_sorted(seq_.begin(), seq_.end())) {
+            util::sort(seq_);
+        }
+        reset();
+    }
+
```

Comment:
I would like to see this either be an `EXPECTS` test, or to make the semantics explicit: that the vector time sequence returns the next time in the vector that is not strictly less than the previous time. The latter has the benefit of being robust to user input without requiring an initial traversal over the whole vector.

Reply:
The later would be more robust to user input, however it would not give what I feel would be the expected behavior for the user.

This is a user facing interface, and I feel that sending in unsorted values is the sort of error we should reasonably expect from users. In which case I am not comfortable modifying our semantics to avoid testing user inputs. I think we just have to do the test.

Reply:
I'd still prefer then to throw than sort, and have the name be e.g. `checked_vector_time_seq` to make it clear. Or else move validation closer to the user, e.g. in `vector_backed_generator` and `spike_source_cell`.